TON Pool Staker

welldan97 opened the pull request:
Add support for Chorus One TON Pool solution, which has 2 pools underneath

mkaczanowski commented on packages/ton/src/TonPoolStaker.ts:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    const payload = beginCell()
+      .storeUint(2077040623, 32)
```
is there any explanation where this uint comes from?

mkaczanowski commented on packages/ton/src/TonPoolStaker.ts:
```diff
@@ -0,0 +1,160 @@
+    const data = await this.getPoolParamsUnformatted({ validatorAddress })
+
+    const payload = beginCell()
+      .storeUint(3665837821, 32)
```
same here

mkaczanowski reviewed (approved):
looks good, apart from minor questions it would be nice to post a few CLI invocations and resulting transactions on testnet to prove code works as expected

nhallenC1 commented:
TON SDK docs look amazing, great work @mkaczanowski and @welldan97! 🚀 

Wanted to raise it here too -- I did notice that the [Signing with Fireblocks](https://github.com/ChorusOne/chorus-one-sdk/blob/main/book/build-your-staking-dapp/signers-explained/fireblocks.md) link appears to be broken.
Upon further review, looks like all the sections that use "Signing with Fireblocks" have that broken link. (For Pool, Single, and Nominator options)
- E.g. the the section [here](https://chorus-one.gitbook.io/sdk/~/revisions/WmXi4yWj1BHJiwtCqOqR/build-your-staking-dapp/ton/ton-pool/overview#signing-the-transaction)

Similarly, each section right above that for "[What is a Signer?](https://github.com/ChorusOne/chorus-one-sdk/blob/main/book/build-your-staking-dapp/signers-explained/what-is-a-signer.md)" right above the Fireblocks section in each pool type have the same issue. 

However, "[What is a Signer?](https://chorus-one.gitbook.io/sdk/~/revisions/WmXi4yWj1BHJiwtCqOqR/signers-explained/what-is-a-signer)" at the bottom of each of those pages works fine! 👍

welldan97 commented:
@mkaczanowski @nhallenC1, I've addressed the comments. I've also added an architecture overview.
Attaching examples of outputs. We should improve those, though I suggest to do in other PR:

### Staking:

https://testnet.tonviewer.com/transaction/c783619915ba10c21e76cb16e804248e1293ac1cd37b579ce9a8163757c1e749
<details>

<summary>Output</summary>

```
> npm run staking-cli -- ton tx delegate-pool 1.3 -c ./config.json -b -s local
# [0/3] inspect input data
{
  "delegator": "0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm",
  "validatorAddress": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
  "messageType": "delegate-pool",
  "args": [
    "1.3"
  ],
  "broadcast": true
}
Delegating to validator #1: kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F
# [1/3] prepare unsigned transaction
{
  "validUntil": 1733257808,
  "message": {
    "address": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
    "bounceable": true,
    "amount": "1300000000",
    "payload": {
      "_hashes": [
        {
          "type": "Buffer",
          "data": [
            43,
            208,
            223,
            254,
            53,
            165,
            0,
            120,
            239,
            123,
            115,
            33,
            122,
            96,
            172,
            60,
            158,
            173,
            243,
            162,
            171,
            24,
            72,
            167,
            235,
            190,
            149,
            194,
            231,
            104,
            36,
            84
          ]
        },
        {
          "type": "Buffer",
          "data": [
            43,
            208,
            223,
            254,
            53,
            165,
            0,
            120,
            239,
            123,
            115,
            33,
            122,
            96,
            172,
            60,
            158,
            173,
            243,
            162,
            171,
            24,
            72,
            167,
            235,
            190,
            149,
            194,
            231,
            104,
            36,
            84
          ]
        },
        {
          "type": "Buffer",
          "data": [
            43,
            208,
            223,
            254,
            53,
            165,
            0,
            120,
            239,
            123,
            115,
            33,
            122,
            96,
            172,
            60,
            158,
            173,
            243,
            162,
            171,
            24,
            72,
            167,
            235,
            190,
            149,
            194,
            231,
            104,
            36,
            84
          ]
        },
        {
          "type": "Buffer",
          "data": [
            43,
            208,
            223,
            254,
            53,
            165,
            0,
            120,
            239,
            123,
            115,
            33,
            122,
            96,
            172,
            60,
            158,
            173,
            243,
            162,
            171,
            24,
            72,
            167,
            235,
            190,
            149,
            194,
            231,
            104,
            36,
            84
          ]
        }
      ],
      "_depths": [
        0,
        0,
        0,
        0
      ],
      "type": -1,
      "bits": {
        "_length": 156,
        "_data": {
          "type": "Buffer",
          "data": [
            123,
            205,
            31,
            239,
            0,
            6,
            205,
            86,
            132,
            244,
            73,
            7,
            48,
            24,
            106,
            0,
            0,
            0,
            83,
            144,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
          ]
        },
        "_offset": 0
      },
      "refs": [],
      "mask": {
        "_mask": 0,
        "_hashIndex": 0,
        "_hashCount": 1
      }
    }
  }
}
Do you want to sign the TX? [y/n]: y
/Users/palestrina/NoBackup/ChorusOne/projects/chorus-one-sdk/packages/ton/src/TonBaseStaker.ts:450
      throw new Error(
            ^


Error: sending 1.3 would result in balance below the minimum existential balance of 0.01 for address 0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm
    at TonPoolStaker.checkMinimumExistentialBalance (/Users/palestrina/NoBackup/ChorusOne/projects/chorus-one-sdk/packages/ton/src/TonBaseStaker.ts:450:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v20.11.0

~/NoBackup/ChorusOne/projects/chorus-one-sdk/packages/staking-cli ton-pool*
> npm run staking-cli -- ton tx delegate-pool 1.2 -c ./config.json -b -s local
# [0/3] inspect input data
{
  "delegator": "0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm",
  "validatorAddress": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
  "messageType": "delegate-pool",
  "args": [
    "1.2"
  ],
  "broadcast": true
}
Delegating to validator #1: kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F
# [1/3] prepare unsigned transaction
{
  "validUntil": 1733257844,
  "message": {
    "address": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
    "bounceable": true,
    "amount": "1200000000",
    "payload": {
      "_hashes": [
        {
          "type": "Buffer",
          "data": [
            160,
            114,
            199,
            255,
            48,
            63,
            112,
            236,
            149,
            200,
            195,
            90,
            121,
            60,
            84,
            70,
            104,
            166,
            11,
            35,
            21,
            20,
            156,
            253,
            105,
            74,
            133,
            3,
            140,
            236,
            235,
            146
          ]
        },
        {
          "type": "Buffer",
          "data": [
            160,
            114,
            199,
            255,
            48,
            63,
            112,
            236,
            149,
            200,
            195,
            90,
            121,
            60,
            84,
            70,
            104,
            166,
            11,
            35,
            21,
            20,
            156,
            253,
            105,
            74,
            133,
            3,
            140,
            236,
            235,
            146
          ]
        },
        {
          "type": "Buffer",
          "data": [
            160,
            114,
            199,
            255,
            48,
            63,
            112,
            236,
            149,
            200,
            195,
            90,
            121,
            60,
            84,
            70,
            104,
            166,
            11,
            35,
            21,
            20,
            156,
            253,
            105,
            74,
            133,
            3,
            140,
            236,
            235,
            146
          ]
        },
        {
          "type": "Buffer",
          "data": [
            160,
            114,
            199,
            255,
            48,
            63,
            112,
            236,
            149,
            200,
            195,
            90,
            121,
            60,
            84,
            70,
            104,
            166,
            11,
            35,
            21,
            20,
            156,
            253,
            105,
            74,
            133,
            3,
            140,
            236,
            235,
            146
          ]
        }
      ],
      "_depths": [
        0,
        0,
        0,
        0
      ],
      "type": -1,
      "bits": {
        "_length": 156,
        "_data": {
          "type": "Buffer",
          "data": [
            123,
            205,
            31,
            239,
            0,
            0,
            4,
            27,
            214,
            35,
            243,
            227,
            48,
            24,
            106,
            0,
            0,
            0,
            83,
            144,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
          ]
        },
        "_offset": 0
      },
      "refs": [],
      "mask": {
        "_mask": 0,
        "_hashIndex": 0,
        "_hashCount": 1
      }
    }
  }
}
Do you want to sign the TX? [y/n]: y
signing data from address: 0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm
* signed transaction received:
{
  tx: x{0100B61BABAD9333BFDA242C578E5FD5F6A6CC11B0E32FD21FEDB33CA3A05F925C92080D7D7D573599C22155D5EB99AE9AEAE6EAC75598ADD359AA11F4E02B0D29A9A317674F6A740000001D0003}
   x{62000382D481C9F89DD45CA6A0CE24876551FF5DD9002D35355E1E7369DF39A6EF95A23C346000000000000000000000000000007BCD1FEF0000041BD623F3E330186A000000539},
  address: '0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm',
  txHash: 'beb3b11d25931749054619db63e16e056cc9a8c415bc4767b54ec132678b8425',
  stateInit: undefined
}
Do you want to broadcast TX? [y/n]: y
# [3/3] broadcasting the signed transaction

Check TX status here: https://testnet.tonviewer.com/transaction/c783619915ba10c21e76cb16e804248e1293ac1cd37b579ce9a8163757c1e749
```

</details>

### Unstaking:

https://testnet.tonviewer.com/transaction/dac72d8231cbc9b6af3d11264d768586fd5066c3825757c984cdd4407a3118fb
<details>

<summary>Output</summary>

```

> npm run staking-cli -- ton tx unstake-pool 2.5 -c ./config.json -b -s local -I 1
# [0/3] inspect input data
{
  "delegator": "0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm",
  "validatorAddress": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
  "messageType": "unstake-pool",
  "args": [
    "2.5"
  ],
  "broadcast": true
}
Unstaking from validator #1: kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F
# [1/3] prepare unsigned transaction
{
  "validUntil": 1733253997,
  "message": {
    "address": "kQAHBakDk_E7qLlNQZxJDsqj_ruyAFpqarw85tO-c03fK26F",
    "bounceable": true,
    "amount": "200000000",
    "payload": {
      "_hashes": [
        {
          "type": "Buffer",
          "data": [
            100,
            102,
            106,
            168,
            3,
            54,
            217,
            83,
            84,
            11,
            236,
            63,
            196,
            198,
            154,
            4,
            123,
            60,
            111,
            234,
            143,
            11,
            41,
            156,
            99,
            169,
            9,
            12,
            212,
            24,
            131,
            205
          ]
        },
        {
          "type": "Buffer",
          "data": [
            100,
            102,
            106,
            168,
            3,
            54,
            217,
            83,
            84,
            11,
            236,
            63,
            196,
            198,
            154,
            4,
            123,
            60,
            111,
            234,
            143,
            11,
            41,
            156,
            99,
            169,
            9,
            12,
            212,
            24,
            131,
            205
          ]
        },
        {
          "type": "Buffer",
          "data": [
            100,
            102,
            106,
            168,
            3,
            54,
            217,
            83,
            84,
            11,
            236,
            63,
            196,
            198,
            154,
            4,
            123,
            60,
            111,
            234,
            143,
            11,
            41,
            156,
            99,
            169,
            9,
            12,
            212,
            24,
            131,
            205
          ]
        },
        {
          "type": "Buffer",
          "data": [
            100,
            102,
            106,
            168,
            3,
            54,
            217,
            83,
            84,
            11,
            236,
            63,
            196,
            198,
            154,
            4,
            123,
            60,
            111,
            234,
            143,
            11,
            41,
            156,
            99,
            169,
            9,
            12,
            212,
            24,
            131,
            205
          ]
        }
      ],
      "_depths": [
        0,
        0,
        0,
        0
      ],
      "type": -1,
      "bits": {
        "_length": 160,
        "_data": {
          "type": "Buffer",
          "data": [
            218,
            128,
            62,
            253,
            0,
            6,
            57,
            34,
            21,
            243,
            90,
            225,
            48,
            24,
            106,
            4,
            149,
            2,
            249,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
          ]
        },
        "_offset": 0
      },
      "refs": [],
      "mask": {
        "_mask": 0,
        "_hashIndex": 0,
        "_hashCount": 1
      }
    }
  }
}
Do you want to sign the TX? [y/n]: y
signing data from address: 0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm
* signed transaction received:
{
  tx: x{503C1E5F0296BE753C01B8D6D0D5F9D309386B9C9A60A339CAF360BD9AD4097E20756D31E9C7D23D4310C3E3749B49981B48E07F18F88897AEA594A9ECB1290929A9A317674F5B6D0000001A0003}
   x{62000382D481C9F89DD45CA6A0CE24876551FF5DD9002D35355E1E7369DF39A6EF95A05F5E100000000000000000000000000000DA803EFD0006392215F35AE130186A049502F900},
  address: '0QDeHoxmpJ6dzzl7KyHDQuOSkwvEnXMcs-8gMCJOSo-IpWcm',
  txHash: '97701a32770ab98e40f83a15be8cb9921721dbd47eb8b7525c031c4742396438',
  stateInit: undefined
}
Do you want to broadcast TX? [y/n]: y
# [3/3] broadcasting the signed transaction

Check TX status here: https://testnet.tonviewer.com/transaction/dac72d8231cbc9b6af3d11264d768586fd5066c3825757c984cdd4407a3118fb


```

</details>
